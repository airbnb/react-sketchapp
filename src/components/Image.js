// @flow
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { or } from 'airbnb-prop-types';
import StyleSheet from '../stylesheet';
import ResizeModePropTypes from './ResizeModePropTypes';
import ImageStylePropTypes from './ImageStylePropTypes';
import { ViewPropTypes } from './View';

const ImageURISourcePropType = PropTypes.shape({
  uri: PropTypes.string.isRequired,
  height: PropTypes.number,
  width: PropTypes.number,
  // bundle: PropTypes.string,
  // method: PropTypes.string,
  // headers: PropTypes.objectOf(PropTypes.string),
  // body: PropTypes.string,
  // cache: PropTypes.oneOf(['default', 'reload', 'force-cache', 'only-if-cached']),
  // scale: PropTypes.number,
});

export const ImageSourcePropType = PropTypes.oneOfType([
  ImageURISourcePropType,
  // PropTypes.arrayOf(ImageURISourcePropType), // TODO: handle me
  PropTypes.string,
]);

const ResizeModes = {
  contain: 'Fit',
  cover: 'Fill',
  stretch: 'Stretch',
  center: 'Fill', // TODO(gold): implement ResizeModes.center
  repeat: 'Tile',
  none: 'Fill',
};

export const ImagePropTypes = {
  ...ViewPropTypes,
  style: or([PropTypes.shape(ImageStylePropTypes), PropTypes.number]),
  defaultSource: ImageSourcePropType,
  resizeMode: ResizeModePropTypes,
  source: ImageSourcePropType,
};

// $FlowFixMe
export default class Image extends React.Component {
  static propTypes = ImagePropTypes;

  static defaultProps = {
    name: 'Image',
  };

  render() {
    const { children, source, defaultSource, resizeMode, name, resizingConstraint } = this.props;

    let style = StyleSheet.flatten(this.props.style) || {};

    const sketchResizeMode = ResizeModes[resizeMode || (style && style.resizeMode) || 'cover'];
    if (source && typeof source !== 'string') {
      style = {
        height: source.height,
        width: source.width,
        ...style,
      };
    }

    return (
      <image
        style={style}
        source={source || defaultSource}
        name={name}
        resizeMode={sketchResizeMode}
        resizingConstraint={resizingConstraint}
      >
        {children}
      </image>
    );
  }
}
