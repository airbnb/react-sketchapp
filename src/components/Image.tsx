import * as React from 'react';
import { FileFormat1 as FileFormat } from '@sketch-hq/sketch-file-format-ts';
import * as PropTypes from 'prop-types';
import { or } from 'airbnb-prop-types';
import StyleSheet from '../stylesheet';
import ResizeModePropTypes from './ResizeModePropTypes';
import ImageStylePropTypes from './ImageStylePropTypes';
import { ViewPropTypes } from './View';

const ImageURISourcePropType = PropTypes.shape({
  uri: PropTypes.string.isRequired,
  height: PropTypes.number,
  width: PropTypes.number,
  // bundle: PropTypes.string,
  // method: PropTypes.string,
  // headers: PropTypes.objectOf(PropTypes.string),
  // body: PropTypes.string,
  // cache: PropTypes.oneOf(['default', 'reload', 'force-cache', 'only-if-cached']),
  // scale: PropTypes.number,
});

export const ImageSourcePropType = PropTypes.oneOfType([
  ImageURISourcePropType,
  // PropTypes.arrayOf(ImageURISourcePropType), // TODO: handle me
  PropTypes.string,
]);

const ResizeModes: { [key: string]: FileFormat.PatternFillType } = {
  contain: 3,
  cover: 1,
  stretch: 2,
  center: 1, // TODO(gold): implement ResizeModes.center
  repeat: 0,
  none: 1,
};

export const ImagePropTypes = {
  ...ViewPropTypes,
  style: or([PropTypes.shape(ImageStylePropTypes), PropTypes.number]),
  defaultSource: ImageSourcePropType,
  resizeMode: ResizeModePropTypes,
  source: ImageSourcePropType,
};

export type Props = PropTypes.InferProps<typeof ImagePropTypes>;

export default class Image extends React.Component<Props> {
  static propTypes = ImagePropTypes;

  static defaultProps = {
    name: 'Image',
  };

  render() {
    const { children, source, defaultSource, resizeMode, name, resizingConstraint } = this.props;

    let style = StyleSheet.flatten(this.props.style) || {};

    const sketchResizeMode = ResizeModes[resizeMode || (style && style.resizeMode) || 'cover'];

    if (source && typeof source !== 'string') {
      style = {
        height: source.height,
        width: source.width,
        ...style,
      };
    }

    return (
      <sketch_image
        style={style}
        source={source || defaultSource}
        name={name}
        resizeMode={sketchResizeMode}
        resizingConstraint={resizingConstraint}
      >
        {children}
      </sketch_image>
    );
  }
}
