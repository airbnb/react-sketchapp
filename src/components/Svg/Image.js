// @flow
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { ImageSourcePropType } from '../Image';
import { numberProp } from './props';

// $FlowFixMe
export default class SVGImage extends React.Component {
  static propTypes = {
    x: numberProp,
    y: numberProp,
    width: numberProp.isRequired,
    height: numberProp.isRequired,
    href: ImageSourcePropType,
    preserveAspectRatio: PropTypes.string,
    children: PropTypes.node,
  };

  static defaultProps = {
    x: 0,
    y: 0,
    width: 0,
    height: 0,
    preserveAspectRatio: 'xMidYMid meet',
  };

  render() {
    const { children, ...rest } = this.props;
    return <svg_image {...rest}>{children}</svg_image>;
  }
}
