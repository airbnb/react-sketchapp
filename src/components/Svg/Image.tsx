import * as React from 'react';
import * as PropTypes from 'prop-types';
import { ImageSourcePropType } from '../Image';
import { numberProp } from './props';

const propTypes = {
  x: numberProp,
  y: numberProp,
  width: numberProp.isRequired,
  height: numberProp.isRequired,
  href: ImageSourcePropType,
  preserveAspectRatio: PropTypes.string,
  children: PropTypes.node,
};

type Props = PropTypes.InferProps<typeof propTypes>;

export class SVGImage extends React.Component<Props> {
  static propTypes = propTypes;

  static defaultProps = {
    x: 0,
    y: 0,
    width: 0,
    height: 0,
    preserveAspectRatio: 'xMidYMid meet',
  };

  render() {
    const { children, ...rest } = this.props;
    return <svg_image {...rest}>{children}</svg_image>;
  }
}
