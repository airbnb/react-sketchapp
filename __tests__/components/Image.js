import React from 'react';
import renderer from 'react-test-renderer';
import Image from '../../src/components/Image';
import StyleSheet from '../../src/stylesheet';

describe('<Image />', () => {
  it('renders children', () => {
    const tree = renderer
      .create(
        <Image source="foo">
          <foo />
        </Image>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('flattens its stylesheet');

  describe('name', () => {
    it('passes its name', () => {
      const tree = renderer.create(<Image source="foo" name="bar" />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('defaults to Image', () => {
      const tree = renderer.create(<Image source="foo" />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('resizeMode', () => {
    it('translates contain', () => {
      const tree = renderer
        .create(<Image source="foo" resizeMode="contain" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('translates cover', () => {
      const tree = renderer
        .create(<Image source="foo" resizeMode="cover" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('translates stretch', () => {
      const tree = renderer
        .create(<Image source="foo" resizeMode="stretch" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('translates center', () => {
      const tree = renderer
        .create(<Image source="foo" resizeMode="center" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('translates repeat', () => {
      const tree = renderer
        .create(<Image source="foo" resizeMode="repeat" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('translates none', () => {
      const tree = renderer
        .create(<Image source="foo" resizeMode="none" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('falls back to cover', () => {
      const tree = renderer.create(<Image source="foo" />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('prefers prop to style', () => {
      const tree = renderer
        .create(
          <Image
            source="foo"
            resizeMode="cover"
            style={{ resizeMode: 'contain' }}
          />
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('falls back to a resizeMode from style', () => {
      const tree = renderer
        .create(<Image source="foo" style={{ resizeMode: 'contain' }} />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('source', () => {
    it('prefers source over defaultSource', () => {
      const tree = renderer
        .create(<Image source="foo" defaultSource="bar" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('falls back to defaultSource if available', () => {
      const tree = renderer.create(<Image defaultSource="foo" />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('sets height from source', () => {
      const tree = renderer
        .create(<Image source={{ uri: 'foo', height: 500 }} />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('sets width from source', () => {
      const tree = renderer
        .create(<Image source={{ uri: 'foo', width: 500 }} />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('favors style over source for height', () => {
      const tree = renderer
        .create(
          <Image
            source={{ uri: 'foo', height: 500 }}
            style={{ height: 400, width: 300 }}
          />
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('favors style over source for width', () => {
      const tree = renderer
        .create(
          <Image
            source={{ uri: 'foo', width: 500 }}
            style={{ height: 400, width: 300 }}
          />
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('style', () => {
    const styles = StyleSheet.create({
      view: {
        flex: 1,
      },
    });

    it('accepts a plain object', () => {
      const tree = renderer.create(<Image style={{ flex: 1 }} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('accepts a StyleSheet ordinal', () => {
      const tree = renderer.create(<Image style={styles.view} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('accepts an array of plain objects and/or StyleSheet ordinals', () => {
      const tree = renderer
        .create(<Image style={[{ flexGrow: 1 }, styles.view]} />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
