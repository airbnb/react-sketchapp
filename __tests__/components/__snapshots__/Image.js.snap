exports[`<Image /> name defaults to Image 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> name passes its name 1`] = `
<image
  name="bar"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> renders children 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}}>
  <foo />
</image>
`;

exports[`<Image /> resizeMode falls back to a resizeMode from style 1`] = `
<image
  name="Image"
  resizeMode="Fit"
  resizingConstraint={undefined}
  source="foo"
  style={
    Object {
      "resizeMode": "contain",
    }
  } />
`;

exports[`<Image /> resizeMode falls back to cover 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> resizeMode prefers prop to style 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source="foo"
  style={
    Object {
      "resizeMode": "contain",
    }
  } />
`;

exports[`<Image /> resizeMode translates center 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> resizeMode translates contain 1`] = `
<image
  name="Image"
  resizeMode="Fit"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> resizeMode translates cover 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> resizeMode translates none 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> resizeMode translates repeat 1`] = `
<image
  name="Image"
  resizeMode="Tile"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> resizeMode translates stretch 1`] = `
<image
  name="Image"
  resizeMode="Stretch"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> source falls back to defaultSource if available 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> source favors style over source for height 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source={
    Object {
      "height": 500,
      "uri": "foo",
    }
  }
  style={
    Object {
      "height": 400,
      "width": 300,
    }
  } />
`;

exports[`<Image /> source favors style over source for width 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source={
    Object {
      "uri": "foo",
      "width": 500,
    }
  }
  style={
    Object {
      "height": 400,
      "width": 300,
    }
  } />
`;

exports[`<Image /> source prefers source over defaultSource 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source="foo"
  style={Object {}} />
`;

exports[`<Image /> source sets height from source 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source={
    Object {
      "height": 500,
      "uri": "foo",
    }
  }
  style={
    Object {
      "height": 500,
      "width": undefined,
    }
  } />
`;

exports[`<Image /> source sets width from source 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source={
    Object {
      "uri": "foo",
      "width": 500,
    }
  }
  style={
    Object {
      "height": undefined,
      "width": 500,
    }
  } />
`;

exports[`<Image /> style accepts a StyleSheet ordinal 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source={undefined}
  style={
    Object {
      "flex": 1,
    }
  } />
`;

exports[`<Image /> style accepts a plain object 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source={undefined}
  style={
    Object {
      "flex": 1,
    }
  } />
`;

exports[`<Image /> style accepts an array of plain objects and/or StyleSheet ordinals 1`] = `
<image
  name="Image"
  resizeMode="Fill"
  resizingConstraint={undefined}
  source={undefined}
  style={
    Object {
      "flex": 1,
      "flexGrow": 1,
    }
  } />
`;
